docs/design/validation.md

```diff
@@ -0,0 +1,1989 @@
+}
+
+func (v stringLengthBetweenValidator) MarkdownDescription(_ context.Context) string {
+    return fmt.Sprintf("length must be between `%d` and `%d`", v.minimum, v.maximum)
+}
+
+func (v stringLengthBetweenValidator) Validate(ctx context.Context, req ValidateAttributeRequest, resp *ValidateAttributeResponse) {
+    value, ok := req.AttributeConfig.(types.String) // see also attr.ValueAs() proposal
+
+    if !ok {
+        resp.Diagnostics = append(resp.Diagnostics, &tfprotov6.Diagnostic{
+            Severity: tfprotov6.DiagnosticSeverityError,
+            Summary: "Invalid value type",
+            Details: fmt.Sprintf("received incorrect value type (%T) at path: %s", req.AttributeConfig, req.Config.AttributePath),
+        })
+        return
+    }
+
+    if req.Config.Unknown {
+        resp.Diagnostics = append(resp.Diagnostics, &tfprotov6.Diagnostic{
+            Severity: tfprotov6.DiagnosticSeverityError,
+            Summary: "Unknown validation value",
+            Details: fmt.Sprintf("received unknown value at path: %s", req.Config.AttributePath),
+        })
+        return
+    }
+
+    if len(req.Config.Value) < v.minimum || len(req.Config.Value) > v.maximum {
+        resp.Diagnostics = append(resp.Diagnostics, &tfprotov6.Diagnostic{
+            Severity: tfprotov6.DiagnosticSeverityError,
+            Summary: "Value validation failed",
+            Details: fmt.Sprintf("%s with value %q %s", req.Config.AttributePath, req.Config.Value, v.Description(ctx))
+        })
+        return
+    }
+
+    return
+}
+
+func StringLengthBetween(minimum int, maximum int) stringLengthBetweenValidator {
+    return stringLengthBetweenValidator{
+        maximum: maximum,
+        minimum: minimum,
+    }
+}
+```
+
+Example provider code:
+
+```go
+schema.Attribute{
+    Type:       types.StringType,
+    Required:   true,
+    Validators: []AttributeValidator{
+        ConflictsWithAttribute(tftypes.NewAttributePath().AttributeName("other_attribute")),
+        StringLengthBetween(1, 256),
+    },
+}
+```
+
+### Future Considerations
+
+It is recommended that the framework provide an abstracted `*tftypes.AttributePath` rather than depend on that type directly, but this can converted after an initial implementation. This is purely for decoupling the two projects, similar to other abstracted types already created in the framework.
+
+To better support provider-based validation functionality in the future, it is also recommended that the `Provider` interface type also add a new `Configured(context.Context) bool` function or another methodology for easily checking the configuration state of a provider instance. Adding a setter function could also allow the framework to manage the provider configuration state automatically. This would simplify validations that require provider instances since it will likely be required that implementations need to check on this status as part of the validation logic.
```

Comment:
I'm 🤔 on this and would be interested in hearing your thoughts. Maybe instead of "recommended", this is something we'd like to further explore and give design consideration to? I just don't want to pre-commit to it being a Thing We Should Do without talking about it a bit more, but I don't know that that conversation needs to happen now or in this design. But I think calling it out as a thing that would impact this that we may want to look into doing is worthwhile.

Reply:
Sure, as is true with all of these. Will update this section's verbiage to recommend potentially discussing these things.